File: GGAndroid_2.0/GGOpen/open-line/src/main/java/io/ganguo/line/share/LineShareService.java
package io.ganguo.line.share;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.provider.MediaStore;
import android.support.annotation.StringDef;

import java.io.File;
import java.io.UnsupportedEncodingException;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.net.URLEncoder;

import io.ganguo.open.sdk.IService;
import io.ganguo.utils.util.AppInstalls;
import io.ganguo.utils.util.Bitmaps;
import io.ganguo.utils.util.Strings;

/**
 * Line 分享
 * Created by leo on 2018/11/8.
 */
public class LineShareService implements IService {
    //Line 路由
    public static String LINE_ROUTE = "line://msg/";
    public static String LINE_PATH = "jp.naver.line.android.activity.selectchat.SelectChatActivity";
    public static String UTF_8 = "UTF-8";

    private Builder builder;

    private LineShareService() {
    }

    /**
     * function：单例  - 懒汉模式
     *
     * @return
     */
    public static LineShareService get() {
        return LineShareService.LazyHolder.HOLDER;
    }


    public static class LazyHolder {
        static LineShareService HOLDER = new LineShareService();
    }

    /**
     * function：添加分享相关参数
     *
     * @param builder
     * @return
     */
    public LineShareService appShareData(Builder builder) {
        this.builder = builder;
        return this;
    }

    @Override
    public boolean apply() {
        try {
            builder.activity.startActivity(createShareTextIntent());
            return true;
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return false;
    }

    @Override
    public void release() {
        builder = null;
    }

    /**
     * function：create line share intent
     *
     * @return
     */
    protected Intent createShareTextIntent() throws UnsupportedEncodingException {
        StringBuilder stringBuilder = new StringBuilder(LINE_ROUTE);
        stringBuilder.append(builder.shareType);
        if (Strings.isNotEmpty(builder.title)) {
            stringBuilder.append(URLEncoder.encode(builder.title + "\n", UTF_8));
        }
        if (Strings.isNotEmpty(builder.content)) {
            stringBuilder.append(URLEncoder.encode(builder.content + "\n", UTF_8));
        }
        if (Strings.isNotEmpty(builder.url)) {
            stringBuilder.append(URLEncoder.encode(builder.url + "\n", UTF_8));
        }
        return new Intent(Intent.ACTION_VIEW, Uri.parse(stringBuilder.toString()));
    }


    /**
     * function：分享类型是否是文本类型
     */
    protected boolean isTextType() {
        if (Strings.isEquals(builder.shareType, LineShareType.LINE_TEXT_TYPE)) {
            return true;
        }
        return false;
    }


    /**
     * function：分享参数
     */
    public static class Builder {
        protected Activity activity;
        protected String content;
        protected String url;
        protected String title;
        @LineShareType
        protected String shareType;

        public Builder(Activity activity, @LineShareType String shareType) {
            this.activity = activity;
            this.shareType = shareType;
        }


        public Builder setTitle(String title) {
            this.title = title;
            return this;
        }

        public Builder setContent(String content) {
            this.content = content;
            return this;
        }

        public Builder setUrl(String url) {
            this.url = url;
            return this;
        }

    }


    @Retention(RetentionPolicy.SOURCE)
    @StringDef(value = {LineShareType.LINE_TEXT_TYPE, LineShareType.LINE_IMAGE_TYPE})
    public @interface LineShareType {
        // TODO: 2018/11/8  暂时只支持分享文本，图片分享研究中
        String LINE_TEXT_TYPE = "text/";
        String LINE_IMAGE_TYPE = "image/";

    }

}

File: GGAndroid_4.0/core/service/service-open/open-line/src/main/kotlin/io/ganguo/line/share/LineShareService.kt
package io.ganguo.line.share

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.provider.MediaStore
import io.ganguo.line.LineHandler
import io.ganguo.line.annotation.LineShareContentType
import io.ganguo.sample.sdk.AShareService
import io.ganguo.sample.sdk.ShareResultObservable
import io.ganguo.sample.sdk.URIs
import io.reactivex.rxjava3.core.Observable
import java.lang.ref.WeakReference
import java.net.URLEncoder


/**
 * <pre>
 *     author : leo
 *     time   : 2019/11/04
 *     desc   : Line分享服务
 * </pre>
 */
@Suppress("DEPRECATION")
class LineShareService(activity: Activity, private var shareEntity: LineShareEntity, private var listener: LineShareListener) : AShareService() {
    private var weakActivity: WeakReference<Activity>? = WeakReference(activity)

    /**
     * 发起分享服务
     * @return ShareResultObservable
     */
    override fun startService(): ShareResultObservable {
        try {
            weakActivity?.get()?.startActivity(buildShareIntent())
            listener.onComplete()
        } catch (e: Throwable) {
            e.printStackTrace()
            return Observable.error(Throwable(e.message.orEmpty(), e))
        }
        return resultObserver
    }

    /**
     * 检测Line SDK配置是否正确
     * @return Throwable?
     */
    override fun checkException(): Throwable? {
        return LineHandler.checkException(weakActivity?.get()!!)
    }


    /**
     * 构建分享Intent
     * @return
     */
    private fun buildShareIntent(): Intent {
        return when (shareEntity.type) {
            LineShareContentType.TYPE_IMAGE -> {
                newShareImageIntent()
            }
            LineShareContentType.TYPE_CONTENT -> {
                newShareContentIntent()
            }
            else -> {
                newShareWebPageIntent()
            }
        }
    }

    /**
     * 创建分享图片Intent
     * @return Intent
     */
    private fun newShareImageIntent(): Intent {
        var uri = Uri.parse(MediaStore.Images.Media.insertImage(weakActivity?.get()?.contentResolver, shareEntity.imageBitmap, null, null))
        val scheme = StringBuilder(LINE_ROUTE)
        scheme.append(formatTypeToIntentType())
        scheme.append(URIs.toAbsolutePath(weakActivity?.get()!!, uri))
        return Intent(Intent.ACTION_VIEW, Uri.parse(scheme.toString()))
    }

    /**
     * 创建分享网页链接Intent
     * @return Intent
     */
    private fun newShareWebPageIntent(): Intent {
        val stringBuilder = StringBuilder(LINE_ROUTE)
        stringBuilder.append(formatTypeToIntentType())
        if (shareEntity.title.isNotEmpty()) {
            stringBuilder.append(URLEncoder.encode(shareEntity.title + "\n", UTF_8))
        }
        if (shareEntity.webPageUrl.isNotEmpty()) {
            stringBuilder.append(URLEncoder.encode(shareEntity.webPageUrl + "\n", UTF_8))
        }
        return Intent(Intent.ACTION_VIEW, Uri.parse(stringBuilder.toString()))

    }

    /**
     * 创建文本分享Intent
     * @return Intent
     */
    private fun newShareContentIntent(): Intent {
        val share = StringBuilder(LINE_ROUTE)
        share.append(formatTypeToIntentType())
        if (shareEntity.title.isNotEmpty()) {
            share.append(URLEncoder.encode(shareEntity.title + "\n", UTF_8))
        }
        if (shareEntity.content.isNotEmpty()) {
            share.append(URLEncoder.encode(shareEntity.content + "\n", UTF_8))
        }
        return Intent(Intent.ACTION_VIEW, Uri.parse(share.toString()))
    }


    /**
     * [LineShareContentType] 转Intent Type
     * @return
     */
    private fun formatTypeToIntentType(): String {
        return when (shareEntity.type) {
            LineShareContentType.TYPE_IMAGE -> {
                "image"
            }
            LineShareContentType.TYPE_CONTENT -> {
                "text/"
            }
            else -> {
                "text/"
            }
        }
    }


    /**
     * @property [LINE_ROUTE] Line路由
     */
    companion object {
        const val LINE_ROUTE = "line://msg/"
        const val UTF_8 = "UTF-8"

    }

    override fun release() {
        super.release()
        weakActivity?.clear()
        weakActivity = null
    }

}